jsapp/js/searches.es6

```diff
@@ -67,50 +51,107 @@ function SearchContext(opts={}) {
       this.listenTo(actions.resources.deleteAsset.completed, this.onDeleteAssetCompleted);
+      this.listenTo(actions.resources.createResource.completed, this.onCreateResourceCompleted);
+      this.listenTo(actions.resources.cloneAsset.completed, this.onCloneAssetCompleted);
+      this.listenTo(actions.resources.loadAsset.completed, this.onLoadAssetCompleted);
+    },
+    onUpdateAssetCompleted(asset) {
+      this.setAsset(asset);
+    },
+    onDeployAssetCompleted(/*data, redeployment, uid*/) {
+      // TODO: deploying asset temporarily triggers loadAsset (already handled
+      // here), so use this event after
+      // https://github.com/kobotoolbox/kpi/issues/1940 is implemented
+    },
+    onDeleteAssetCompleted(asset) {
+      this.removeAsset(asset.uid);
+    },
+    onCreateResourceCompleted(asset) {
+      this.setAsset(asset);
+    },
+    onCloneAssetCompleted(asset) {
+      this.setAsset(asset);
     },
-    onDeleteAssetCompleted (asset) {
-      var filterOutDeletedAsset = ({listName}) => {
-        if (this.state[listName] != undefined) {
-          let uid = asset.uid;
-          let listLength = this.state[listName].length;
-          let l = this.state[listName].filter(function(result){
-            return result.uid !== uid;
-          });
-          if (l.length !== listLength) {
-            let o = {};
-            o[listName] = l;
-            this.update(o);
+    onLoadAssetCompleted(asset) {
+      this.setAsset(asset);
+    },
+    // add/update asset in all search store lists
+    setAsset(asset) {
+      this.setAssetInList(asset, 'defaultQueryResultsList');
+      this.setAssetInList(asset, 'searchResultsList');
+      this.rebuildCategorizedList(
+        this.state.defaultQueryResultsList,
+        'defaultQueryCategorizedResultsLists'
+      );
+      this.rebuildCategorizedList(
+        this.state.searchResultsList,
+        'searchResultsCategorizedResultsLists'
+      );
+    },
+    setAssetInList(asset, listName) {
+      const list = this.state[listName];
+      if (list && list.length !== 0) {
+        const updateObj = {};
+        let isNewAsset = true;
+        for (let i = 0; i < list.length; i++) {
+          if (list[i].uid === asset.uid) {
+            list[i] = asset;
+            isNewAsset = false;
+            break;
           }
         }
-      };
-      var filterOutDeletedAssetFromCategorizedList = () => {
-        let list = this.state.defaultQueryCategorizedResultsLists;
-        if (list) {
-          var l = {};
-          for (var category in list) {
-            l[category] = list[category].filter(function(result){
-              return result.uid !== asset.uid;
-            });
-          }
-          let o = {};
-          o.defaultQueryCategorizedResultsLists = l;
-          this.update(o);
+        if (isNewAsset) {
+          // we add asset on top, because it's update event (freshly modified)
```

Comment:
What's the scenarion that this newAsset case is triggered? I would like to test it.

Reply:
This is definitely being called after cloning an asset and possibly some other actions - i.e. all the times when given asset did not exist in searches yet.